src/cdef.rs

```diff
@@ -74,26 +84,38 @@ impl cdef::Fn {
         let dst_ptr = dst.as_mut_ptr::<BD>().cast();
         let stride = dst.stride();
         let left = ptr::from_ref(left).cast();
-        let top = top.cast();
-        let bottom = bottom.cast();
+        let top_ptr = (&*top.element_as(top_off) as *const BD::Pixel).cast();
```

Comment:
Provenance wise, `.element_as` is incorrect here since it will turn into a slice in the fallbacks, and the bounds check is only checking that one element.  What I did for `.as_ptr` for `Rav1dPictureDataComponentOffset` is just check that the offset is `<=` the length, corresponding to slicing/`.add` being safe.  This is called `.as_ptr_at` (and a `mut` variant) for `Rav1dPictureDataComponent`, so we could add it to `DisjointMut`, too, I think.  Internally, it calls `.as_mut_ptr().add(offset)` after checking `offset` is in bounds (and see that `fn` for how the out of bounds path should be a non-generic `#[inline(never)]` call).

Reply:
> Provenance wise, `.element_as` is incorrect here since it will turn into a slice in the fallbacks, and the bounds check is only checking that one element.

I don't understand what you mean by "turn into a slice in the fallback" here. We're not using the top pointer to construct a slice in the fallback, we use the top buffer directly. So any index operations in the fallback will be bounds checked. So I think `element_as` is reasonable here because it handles both the bounds check and using pixel offsets.

Reply:
Oh, true, you're only using this to recalculate the offset, so provenance-wise it's fine.  That said, I think the check should be `<= len` (`.slice_as`) rather than `< len` (`.element_as`).

Reply:
That doesn't really make sense to me, either. `slice_as` requires a range to slice, but in this case we don't know the range up front, we only know the initial offset. Also I don't see how `element_as`'s bounds check would be different from `slice_as`? Don't both require that `offset < len`?

Reply:
Sorry, I think I got a bit confused here.  Forget about the slicing.

What `Rav1dPictureDataComponentOffset::as_ptr` and `Rav1dPictureDataComponent::as_ptr_at` do is effectively a checked `.add` on the raw ptr.  No intermediary reference or guard is created.  This is provably safe to do because we check the bounds, and thus `.add` is safe.  This makes no guarantees on how the ptr will be used later, which is correct, other than that it is in bounds of course.

In contrast, using something like `.element_as` creates an intermediary guard and reference that we then cast to a ptr and drop.  This has very different semantics.  The creation of this intermediary `&` asserts that there is no other `&mut` concurrently, which we do not know, because we don't know (in general, though we may look at `fn`s case-by-case) that uses of this ptr, or the `DisjointMut` in the fallback, will access this element.  They may first offset it by some amount and access different elements, while another thread may access that first element.  That would be UB.

Reply:
This is done now that I'm using `PicOrBuf`.